core/src/queue.rs

```diff
@@ -205,53 +213,57 @@ impl Queue {
         &self,
         seen: &mut Vec<HashOf<VersionedSignedTransaction>>,
         wsv: &WorldStateView,
+        expired_transactions: &mut Vec<VersionedAcceptedTransaction>,
     ) -> Option<VersionedAcceptedTransaction> {
-        loop {
-            let hash = self.signature_buffer.pop()?;
-            let entry = match self.txs.entry(hash) {
-                Entry::Occupied(entry) => entry,
-                // FIXME: Reachable under high load. Investigate, see if it's a problem.
-                Entry::Vacant(_) => continue,
-            };
-
-            match self.check_tx(entry.get(), wsv) {
-                Ok(MustUse(signature_check)) => {
-                    // Transactions are not removed from the queue until expired or committed
-                    seen.push(hash);
-                    if signature_check {
-                        return Some(entry.get().clone());
-                    }
-                }
-                Err(_) => {
-                    entry.remove_entry();
-                }
-            }
-        }
+        // NOTE: `SKIP_SIGNATURE_CHECK=false` because `signature_buffer` contains transaction which signature check can be either `true` or `false`.
+        self.pop_from::<false>(&self.signature_buffer, seen, wsv, expired_transactions)
     }
 
     /// Pop single transaction from the queue. Record all visited and not removed transactions in `seen`.
     fn pop_from_queue(
         &self,
         seen: &mut Vec<HashOf<VersionedSignedTransaction>>,
         wsv: &WorldStateView,
+        expired_transactions: &mut Vec<VersionedAcceptedTransaction>,
+    ) -> Option<VersionedAcceptedTransaction> {
+        // NOTE: `SKIP_SIGNATURE_CHECK=true` because `queue` contains only transactions for which signature check is `true`.
+        self.pop_from::<true>(&self.queue, seen, wsv, expired_transactions)
+    }
+
+    /// Pop single transaction either from the queue or waiting buffer
+    #[inline]
+    fn pop_from<const SKIP_SIGNATURE_CHECK: bool>(
+        &self,
+        queue: &ArrayQueue<HashOf<VersionedSignedTransaction>>,
+        seen: &mut Vec<HashOf<VersionedSignedTransaction>>,
+        wsv: &WorldStateView,
+        expired_transactions: &mut Vec<VersionedAcceptedTransaction>,
```

Comment:
Not sure if it makes sense or I'm being naive here, but since all we end up using is hash and ttl, maybe it makes sense not to retain the full struct here?

Reply:
Good point, i considered such approach, but decided to leave queue a bit more general.

Reply:
I'm mostly worried this might impact performance. Let's see if anyone performance-savvy has input on this, otherwise I'll try and look into it myself

Reply:
Why are you asking, for someone? Benchmark.

Reply:
@QuentinI @appetrosyan  benchmarking results don't show any performance degradation. 

**NOTE**: Since we are measuring tps and not time results are inverted ("+" is "-", regressed is improved).  

```
Transaction TTL=u64::MAX

Before (aec7aaa5):
time:   [1479.7 tps 1588.9 tps 1693.2 tps]

After:
time:   [1354.4 tps 1595.6 tps 1812.5 tps]
change: [-15.814% +0.4261% +16.030%] (p = 0.96 > 0.05)
No change in performance detected.

Transaction TTL=1_000

Before (aec7aaa5):
# Most of the transactions just expired
time:   [159.19 tps 191.36 tps 223.89 tps] 

After:
time:   [227.29 tps 250.73 tps 273.30 tps]
change: [+9.1402% +31.029% +60.827%] (p = 0.01 < 0.05)
Performance has regressed.
```